### PR TITLE
feat: limit the number of denoms in a plan

### PR DESCRIPTION
**Pull request by hallazzang**
## Description

Closes: #256 

## Tasks

- [x] Add `PrivatePlanMaxNumDenoms`, `PublicPlanMaxNumDenoms` params
- [x] Add `RewardDenoms` field to `RatioPlan`
- [x] Limit the number of coin denoms in a plan
- [x] Check if coins specified in a plan has positive supply
- [x] Update tutorial docs (maybe after bumping up the module version -> v1.1.0 )

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Appropriate labels applied
- [x] Targeted PR against correct branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://go.dev/blog/godoc).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by dongsam** [comment]
@hallazzang How about rename `PrivatePlanMaxNumDenoms`,  `PublicPlanMaxNumDenoms` to `PrivatePlanMaxEntries`, `PublicPlanMaxEntries` to cover `StakingCoinWeights`, `EpochAmount` and `RewardDenoms`

ref. `[stakingparams.MaxEntries](https://github.com/cosmos/cosmos-sdk/blob/v0.44.6/x/staking/types/staking.pb.go#L842)` of sdk `x/staking` also cover unbending and redelegation both

File: x/farming/client/cli/utils.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ type PrivateRatioPlanRequest struct {
 	StartTime          time.Time    `json:"start_time"`
 	EndTime            time.Time    `json:"end_time"`
 	EpochRatio         sdk.Dec      `json:"epoch_ratio"`
+	RewardDenoms       []string     `json:"reward_denoms"`
```
Comment (kogisin) [inline x/farming/client/cli/utils.go:30]: Can you add this param in `TestParsePrivateRatioPlan` case?

File: x/farming/types/params.go [hunk 2]
```diff
@@ -10,6 +10,15 @@ import (
...
 
+const (
+	// PrivatePlanMaxNumDenoms is the maximum number of denoms in a private plan's
+	// staking coin weights, epoch amount and reward denoms.
+	PrivatePlanMaxNumDenoms = 50
+	// PublicPlanMaxNumDenoms is the maximum number of denoms in a public plan's
+	// staking coin weights, epoch amount and reward denoms
+	PublicPlanMaxNumDenoms = 500
+)
+
```
Comment (kogisin) [inline x/farming/types/params.go:21]: Don't we need to specific these variables somewhere in docs?

**Review by kogisin — CHANGES_REQUESTED** [review]

File: x/farming/client/cli/utils.go (hunk 1 above)
Comment (hallazzang) [reply x/farming/client/cli/utils.go:30]: Nice catch. I added that check.

File: x/farming/types/params.go (hunk 2 above)
Comment (hallazzang) [reply x/farming/types/params.go:21]: These constants are for extreme cases and usually users won't even notice that these limits present.
Do you have any idea where to put these? I'm considering creating a new section 'Global constants` in `spec/07_params.md`.

Comment (kogisin) [reply x/farming/types/params.go:21]: I thought the same. Having global constants in `07_params.md` sounds good to me. However, if these variables don't necessarily need to be exposed then it is fine with me not to have them in docs.

Comment (hallazzang) [reply x/farming/types/params.go:21]: I added the documentation about those constants.

**Review by kogisin — APPROVED** [review]

**Comment by dongsam** [comment]
As a result of internal discussions, it was decided to include `RatioPlan` functionality after long-term review and correction without solving the issue by adding 'RewardDenoms' to RatioPlan. It is planned to prevent the creation of `RatioPlan` through separate PR.

**Pull request closed without merging**